### PR TITLE
Add dialog to notebooks for managing Pip packages

### PR DESCRIPTION
**Pull request by corivera**
image

image

### CODE REVIEW

File: extensions/notebook/package.nls.json [hunk 1]
```diff
@@ -24,5 +24,5 @@
 	"config.jupyter.kernelConfigValuesDescription": "Configuration options for Jupyter kernels. This is automatically managed and not recommended to be manually edited.",
 	"title.reinstallNotebookDependencies": "Reinstall Notebook dependencies",
 	"title.configurePython": "Configure Python for Notebooks",
-	"title.installPackages": "Install Packages"
+	"title.managePackages": "Manage Pip Packages"
```
Comment (kevcunnane) [inline extensions/notebook/package.nls.json]: Please run this by PM / UX. Looks good to me but doesn't hurt to ask.

File: extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+	private packagesSearchButton: azdata.ButtonComponent;
+	private newPackagesName: azdata.TextComponent;
+	private newPackagesNameLoader: azdata.LoadingComponent;
+	private newPackagesVersions: azdata.DropDownComponent;
+	private newPackagesVersionsLoader: azdata.LoadingComponent;
+	private newPackagesSummary: azdata.TextComponent;
+	private newPackagesSummaryLoader: azdata.LoadingComponent;
+	private packageInstallButton: azdata.ButtonComponent;
+
+	private readonly InstallButtonText = localize('managePackages.installButtonText', "Install");
```
Comment (kevcunnane) [inline extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts]: super minor: I think elsewhere we stick these as const values outside the class, or else just inline them. Up to you, would be nice to be consistent across our stack.

File: extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+	private readonly AddNewTabTitle = localize('managePackages.addNewTabTitle', "Add new");
+	private readonly InvalidTextPlaceholder = localize('managePackages.invalidTextPlaceholder', "N/A");
+
+	private readonly SearchBarPlaceholder = localize('managePackages.searchBarPlaceholder', "Search for packages");
+	private readonly SearchButtonLabel = localize('managePackages.searchButtonLabel', "Search");
+	private readonly PackageNameTitle = localize('managePackages.packageNameTitle', "Package Name");
+	private readonly PackageVersionTitle = localize('managePackages.packageVersionTitle', "Package Version");
+	private readonly PackageSummaryTitle = localize('managePackages.packageSummaryTitle', "Package Summary");
+
+	private readonly PackageNotFoundError = localize('managePackages.packageNotFound', "Could not find the specified package.");
```
Comment (kevcunnane) [inline extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts]: Nitpick: For errors do we need a `.` at the end? Where does this show up?

File: extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+			let formModel = view.modelBuilder.formContainer()
+				.withFormItems([{
+					component: this.newPackagesSearchBar,
+					title: ''
+				}, {
+					component: this.packagesSearchButton,
+					title: ''
+				}, {
+					component: this.newPackagesNameLoader,
+					title: this.PackageNameTitle
```
Comment (kevcunnane) [inline extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts]: I think you can just inline the localize calls here, that's what we do elsewhere in the code to tighten the classes up.

File: extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+				value: this.InvalidTextPlaceholder
+			});
+		} finally {
+			await this.toggleNewPackagesFields(true);
+		}
+	}
+
+	private async fetchPypiPackage(packageName: string): Promise<PipPackageData> {
+		return new Promise<PipPackageData>((resolve, reject) => {
+			request.get(`https://pypi.org/pypi/${packageName}/json`, { timeout: 10000 }, (error, response, body) => {
```
Comment (kevcunnane) [inline extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts]: So... this won't work behind a firewall. Can you open an issue to track the need to put this into a user setting? That way they could override pypy.org with their own org URL and things will still work. No need to do that in this PR but would be a nice follow up given so many of our users have this kind of issue.

File: extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+				if (error) {
+					return reject(error);
+				}
+
+				if (response.statusCode === 404) {
+					return reject(this.PackageNotFoundError);
+				}
+
+				if (response.statusCode !== 200) {
+					return reject(response.statusMessage);
```
Comment (kevcunnane) [inline extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts]: Is this wrapped upstream? E.g. will we just give a `FORBIDDEN` error message if we get a 401, or would we have a user-friendly message to wrap it?

File: extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+				}
+
+				let packagesJson = JSON.parse(body);
+				let versionKeys = Object.keys(packagesJson.releases);
+				versionKeys = versionKeys.filter(versionKey => {
```
Comment (kevcunnane) [inline extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts]: minor: should we null check?

File: extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+					let releaseInfo = packagesJson.releases[versionKey];
+					return Array.isArray(releaseInfo) && releaseInfo.length > 0;
+				});
+				versionKeys.sort((first, second) => {
+					// sort in descending order
+					let firstVersion = first.split('.').map(numStr => Number.parseInt(numStr));
+					let secondVersion = second.split('.').map(numStr => Number.parseInt(numStr));
+
+					// If versions have different lengths, then append zeroes to the shorter one
+					if (firstVersion.length > secondVersion.length) {
```
Comment (kevcunnane) [inline extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts]: While your logic looks great, could we just use the `semver` npm package instead? It has `.gt` and `.lt` comparators built in. We already use it in other places in Azure Data Studio.

Comment (kevcunnane) [reply extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts]: https://www.npmjs.com/package/semver

File: extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+				resolve({
+					name: packageName,
+					versions: versionKeys,
+					summary: packagesJson.info.summary
+				});
+			});
+		});
+	}
+
+	private async doPackageInstall(): Promise<void> {
```
Comment (kevcunnane) [inline extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts]: Could we wrap this in the azdata tasks API? At some point this should have support for progress messages etc., and that's the standard way to do a potentially long-running task like this.

File: extensions/notebook/src/dialog/managePackages/installedPackagesTab.ts [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+	public async loadInstalledPackagesInfo(): Promise<void> {
+		let packageData: string[][];
+		await this.installedPackagesLoader.updateProperties({ loading: true });
+		try {
+			let pythonPackages = await this.jupyterInstallation.getInstalledPipPackages();
+			let packagesLocation = await this.jupyterInstallation.getPythonPackagesPath();
+
+			await this.installedPackageCount.updateProperties({
+				value: localize('managePackages.packageCountLabel', "{0} packages found in '{1}'",
+					pythonPackages.length,
```
Comment (kevcunnane) [inline extensions/notebook/src/dialog/managePackages/installedPackagesTab.ts]: Do we need to worry about undefined here?

File: extensions/notebook/src/dialog/managePackages/installedPackagesTab.ts [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+	public get tab(): azdata.window.DialogTab {
+		return this.installedPkgTab;
+	}
+
+	public async loadInstalledPackagesInfo(): Promise<void> {
+		let packageData: string[][];
+		await this.installedPackagesLoader.updateProperties({ loading: true });
+		try {
+			let pythonPackages = await this.jupyterInstallation.getInstalledPipPackages();
```
Comment (kevcunnane) [inline extensions/notebook/src/dialog/managePackages/installedPackagesTab.ts]: Can this ever appear if the install isn't complete? Do we need to handle that case? I'm likely just being paranoid 😄

File: extensions/notebook/src/dialog/managePackages/installedPackagesTab.ts [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+					return {
+						name: row[0],
+						version: row[1]
+					};
+				});
+
+				let packagesStr = packages.map(pkg => {
+					return `${pkg.name} ${pkg.version}`;
+				}).join(', ');
+				this.dialog.showInfoMessage(
```
Comment (kevcunnane) [inline extensions/notebook/src/dialog/managePackages/installedPackagesTab.ts]: Again, please wrap in a task

File: extensions/notebook/src/integrationTest/notebookIntegration.test.ts [hunk 13]
```diff
@@ -80,4 +80,28 @@ describe('Notebook Extension Python Installation', function () {
...
+
+		let packagePath = await install.getPythonPackagesPath();
+		should(packagePath).not.be.undefined();
+		should(packagePath.length).be.greaterThan(0);
+
+		let testPkg = 'pandas';
+		let testPkgVersion = '0.24.2';
+		let expectedPkg: PythonPkgDetails = { name: testPkg, version: testPkgVersion };
+
+		await install.installPipPackage(testPkg, testPkgVersion);
```
Comment (kevcunnane) [inline extensions/notebook/src/integrationTest/notebookIntegration.test.ts]: Tests! Awesome!

File: extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts (hunk 8 above)
Comment (corivera) [reply extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts]: Looks like semver fails on versions with less than 3 parts. "0.1" is an invalid version, for example.

File: extensions/notebook/src/dialog/managePackages/installedPackagesTab.ts (hunk 11 above)
Comment (corivera) [reply extensions/notebook/src/dialog/managePackages/installedPackagesTab.ts]: Currently the Manage Packages button appears as enabled during install. I've opened a separate issue for it here: https://github.com/microsoft/azuredatastudio/issues/5955

File: extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts (hunk 5 above)
Comment (corivera) [reply extensions/notebook/src/dialog/managePackages/addNewPackageTab.ts]: https://github.com/microsoft/azuredatastudio/issues/5957

**Pull request merged**